components/utils.py

```diff
@@ -310,26 +336,72 @@ def yum_install(source):
         source_name = subprocess.check_output(
             ['rpm', '-qp', source_path]).strip()
 
-    ctx.logger.info('Checking whether {0} is already installed...'.format(
-        source_path.rstrip('\n\r')))
-    installed = run(['rpm', '-q', source_name], ignore_failures=True)
-    if installed.returncode == 0:
-        ctx.logger.info('Package {0} is already installed.'.format(source))
-        return
+        ctx.logger.info('Checking whether {0} is already installed...'.format(
+            source_path))
+        installed_rpm = _get_installed_rpm_source(source_name)
+        if installed_rpm == source_name:
+            ctx.logger.info('Package {0} is already installed.'.format(source))
+            return
+
+        if is_upgrade() or is_rollback():
+            _remove_existing_rpm_package(source_name)
 
     ctx.logger.info('yum installing {0}...'.format(source_path))
     sudo(['yum', 'install', '-y', source_path])
 
 
+def _remove_existing_rpm_package(source_name):
+    package_name, package_version = _get_rpm_package_details(source_name)
+    existing_src_name = _get_installed_rpm_source(package_name)
+    if existing_src_name:
+        existing_package_name, existing_pkg_version = \
+            _get_rpm_package_details(existing_src_name)
+        ctx.logger.info('Version {0} is currently installed for package {1}. '
+                        'Uninstalling to allow upgrade to version {2}'
+                        .format(existing_pkg_version, existing_src_name,
+                                package_version))
+        sudo(['rpm', '--noscripts', '-e', existing_src_name])
+
+
+def _get_installed_rpm_source(name):
+    '''
```

Comment:
why do we need to return the `existing_rpm_pkg_name` here? We must know the formal name beforehand to even check if it's installed in the first place? `yum_install` implemented that as well in L336